### PR TITLE
[MIRROR] Adds Toggleable Stomachs w/ toggleable visibility

### PR DESCRIPTION
**Pull request by CHOMPStation2**
Original PR: https://github.com/VOREStation/VOREStation/pull/16052
--------------------
Ports Stomach Sprites from Rogue Star, which was originally made on Chomp. Credit for the original code goes to Chomp. Credit for the sprites goes to Chomp.

FEATURES PORTED:
### - Taur halves of participating species (Wolf, Naga, Horse, Cow, Lizard, SynthLizard, Feline, Slug, Drake, Otie, Deer, and Skunk) are able to have stomach sprites on their taur half. (Note: It appears as though the skunk spritee does not exist, yet.)

Porting the above took approximately 3 workhours.

FEATURES CREATED:

### - Adds a toggle for stomach visibility clientside. (This allows you to make it so you can/can't see stomachs depending on your preferences.)
- Said toggle can be found in two places: The Character-setup vore tab and the preferences tab in game. This can be toggled on and off at will.

Notes:
- Photos will not show stomachs.
- Any /mob can toggle the preference on or off. Includes ghosts, AI, borgs, etc.

The above feature took over six hours to properly code. It involded planes/vision scapes which I have never worked with, personally, so I spent a while screwing around trying to figure out how to make it a HUD to no avail. Thanks to @ Heroman3003 I was able to condense my spaghetti code down into a singular line of code and get it functional.

Photos:
Naga Stomach:
image

Otie Stomach:
image

Normal Stomachs:
image

### CODE REVIEW

File: code/modules/mob/living/carbon/human/human.dm [hunk 1]
```diff
@@ -7,6 +7,17 @@
...
+
+	var/vore_capacity = 3
+	var/vore_capacity_ex = list("stomach" = 3, "taur belly" = 3)
+	var/vore_fullness_ex = list("stomach" = 0, "taur belly" = 0)
+	var/vore_icon_bellies = list("stomach", "taur belly")
+	var/struggle_anim_stomach = FALSE
+	var/struggle_anim_taur = FALSE
+	var/vore_sprite_color = list("stomach" = "#FFFFFF", "taur belly" = "#FFFFFF")
+	var/vore_sprite_multiply = list("stomach" = TRUE, "taur belly" = TRUE)
+	var/vore_fullness = 0
```
Comment (Kashargul) [inline code/modules/mob/living/carbon/human/human.dm:19]: Moved from mob to human, comment that section out...

File: code/modules/mob/living/carbon/human/human.dm [hunk 2]
```diff
@@ -1789,3 +1800,19 @@
...
+	for(var/obj/belly/B as anything in vore_organs)
+		new_fullness[B.belly_sprite_to_affect] += B.GetFullnessFromBelly()
+	for(var/belly_class in vore_icon_bellies)
+		new_fullness[belly_class] /= size_multiplier //Divided by pred's size so a macro mob won't get macro belly from a regular prey.
+		new_fullness[belly_class] = round(new_fullness[belly_class], 1) // Because intervals of 0.25 are going to make sprite artists cry.
+		vore_fullness_ex[belly_class] = min(vore_capacity_ex[belly_class], new_fullness[belly_class])
+		vore_fullness += new_fullness[belly_class]
+	vore_fullness = min(vore_capacity, vore_fullness)
+	update_vore_belly_sprite()
+	update_vore_tail_sprite()
```
Comment (Kashargul) [inline code/modules/mob/living/carbon/human/human.dm:1818]: moved from mob to human, compare with our proc override on human... likely comment out as well

File: code/modules/mob/living/carbon/human/species/species.dm [hunk 3]
```diff
@@ -250,6 +250,8 @@
 	var/rarity_value = 1									// Relative rarity/collector value for this species.
 	var/economic_modifier = 2								// How much money this species makes
 
+	var/vore_belly_default_variant = "H"
+
```
Comment (Kashargul) [inline code/modules/mob/living/carbon/human/species/species.dm:254]: moved from our modular_chomp file. remove it from our modular file

File: code/modules/mob/living/carbon/human/update_icons.dm [hunk 4]
```diff
@@ -105,6 +106,48 @@ GLOBAL_LIST_EMPTY(damage_icon_parts) //see UpdateDamageIcon()
...
+#define TAIL_UPPER_LAYER_ALT	33		//Modified tail-sprite layer. Tend to be larger.
+#define MODIFIER_EFFECTS_LAYER	34		//Effects drawn by modifiers
+#define FIRE_LAYER				35		//'Mob on fire' overlay layer
+// # define MOB_WATER_LAYER			36		//'Mob submerged' overlay layer // Moved to global defines
+#define TARGETED_LAYER			37		//'Aimed at' overlay layer
+#define VORE_BELLY_LAYER		38
+#define VORE_TAIL_LAYER			39
+
+#define TOTAL_LAYERS			39		//VOREStation edit. <---- KEEP THIS UPDATED, should always equal the highest number here, used to initialize a list.
+>>>>>>> 8ff0932889... Adds Toggleable Stomachs w/ toggleable visibility (#16052)
```
Comment (Kashargul) [inline code/modules/mob/living/carbon/human/update_icons.dm:150]: Check if the new define is in our globals.... if so, just add it here, but keep it commented out.

File: code/modules/mob/living/carbon/human/update_icons.dm [hunk 5]
```diff
@@ -1447,6 +1490,74 @@ GLOBAL_LIST_EMPTY(damage_icon_parts) //see UpdateDamageIcon()
...
+	return null
+
+/mob/living/carbon/human/proc/vore_tail_animation()
+	if(tail_style.struggle_anim && !struggle_anim_taur)
+		struggle_anim_taur = TRUE
+		update_vore_tail_sprite()
+		spawn(12)
+			struggle_anim_taur = FALSE
+			update_vore_tail_sprite()
+
```
Comment (Kashargul) [inline code/modules/mob/living/carbon/human/update_icons.dm:1560]: copy over our proc version from modular_chomp, remove in modular_chomp and add chompedits here

File: code/modules/mob/new_player/sprite_accessories_taur.dm [hunk 6]
```diff
@@ -356,3 +363,72 @@
...
+	belly_variant_when_loaf = TRUE
+	fullness_icons = 1
+	struggle_anim = TRUE
+
+/datum/sprite_accessory/tail/taur/skunk
+	vore_tail_sprite_variant = "Skunk"
+	belly_variant_when_loaf = TRUE
+	fullness_icons = 1
+	struggle_anim = TRUE
+>>>>>>> 8ff0932889... Adds Toggleable Stomachs w/ toggleable visibility (#16052)
```
Comment (Kashargul) [inline code/modules/mob/new_player/sprite_accessories_taur.dm:434]: likely also copied from the modular_chomp, same as comment above

File: code/modules/vore/eating/belly_obj_vr.dm [hunk 7]
```diff
@@ -64,6 +70,30 @@
...
+	var/health_impacts_size = TRUE
+	var/resist_triggers_animation = TRUE
+	var/size_factor_for_sprite = 1
+	var/belly_sprite_to_affect = "stomach"
+	var/datum/sprite_accessory/tail/tail_to_change_to = FALSE
+	var/tail_colouration = FALSE
+	var/tail_extra_overlay = FALSE
+	var/tail_extra_overlay2 = FALSE
+	var/undergarment_chosen = "Underwear, bottom"
+
```
Comment (Kashargul) [inline code/modules/vore/eating/belly_obj_vr.dm:96]: likely also vars copied over from belly_obj_ch, best to check

File: code/modules/vore/eating/belly_obj_vr.dm [hunk 8]
```diff
@@ -2845,3 +2922,38 @@
...
+			else if(I.w_class == ITEMSIZE_LARGE)
+				fullness_to_add = ITEMSIZE_COST_LARGE
+			else if(I.w_class == ITEMSIZE_HUGE)
+				fullness_to_add = ITEMSIZE_COST_HUGE
+			else
+				fullness_to_add = ITEMSIZE_COST_NO_CONTAINER
+			fullness_to_add /= 32
+			belly_fullness += fullness_to_add * item_multiplier
+	belly_fullness *= size_factor_for_sprite
+	return belly_fullness
```
Comment (Kashargul) [inline code/modules/vore/eating/belly_obj_vr.dm:2959]: proc copied from belly_obj_ch, copy over, add chompedit comments if needed

File: code/modules/vore/eating/vorepanel_vr.dm [hunk 9]
```diff
@@ -3999,6 +4035,101 @@ var/global/list/belly_colorable_only_fullscreens = list("a_synth_flesh_mono",
...
+					hhost.vore_selected.tail_extra_overlay = newcolor
+				. = TRUE
+		if("b_tail_color3")
+			if (istype(host, /mob/living/carbon/human))
+				var/mob/living/carbon/human/hhost = host
+				var/newcolor = input(usr, "Choose tail tertiary color.", "", hhost.vore_selected.tail_extra_overlay2) as color|null
+				if(newcolor)
+					hhost.vore_selected.tail_extra_overlay2 = newcolor
+				. = TRUE
+>>>>>>> 8ff0932889... Adds Toggleable Stomachs w/ toggleable visibility (#16052)
```
Comment (Kashargul) [inline code/modules/vore/eating/vorepanel_vr.dm:4132]: deconf, we likely already have them, so they might be twice now

**Review by Kashargul — CHANGES_REQUESTED** [review]

**Comment by Verkister** [comment]
Didn't even port the fixed lizard belly size 1 sprites

**Comment by Verkister** [comment]
Also why are the belly sprites in the preview pics using fuzzy scaling while the character is not?

**Comment by Kashargul** [comment]
> Also why are the belly sprites in the preview pics using fuzzy scaling while the character is not?

Don't even ask it's a cross backport from rogue from what I see.... I'll try to find some time to deconf it... I've marked the sections with comments where I remembered before sleep

File: code/modules/mob/new_player/sprite_accessories_taur.dm (hunk 6 above)
Comment (Verkister) [reply code/modules/mob/new_player/sprite_accessories_taur.dm:434]: A good chunk of these don't even have struggle anim sprites in the dmi despite being set to TRUE here. Gonna have some invisible guts. Only "N" "E" and "SectDrone" versions have anims

Comment (Verkister) [reply code/modules/mob/new_player/sprite_accessories_taur.dm:434]: Same with the _vr file below.

File: code/modules/mob/new_player/sprite_accessories_taur_vr.dm [hunk 10]
```diff
@@ -165,34 +172,43 @@
 	msg_prey_stepunder = "You jump over %prey's thick tail."
 	msg_owner_stepunder = "%owner bounds over your tail."
 
-/datum/sprite_accessory/tail/taur/naga/naga_2c
+/datum/sprite_accessory/tail/taur/naga/vore_compatable
```
Comment (Verkister) [inline code/modules/mob/new_player/sprite_accessories_taur_vr.dm:175]: Do these path changes have some importance here?

Comment (Verkister) [reply code/modules/mob/new_player/sprite_accessories_taur_vr.dm:175]: also it's spelled compatible smh

File: code/modules/vore/eating/vorepanel_vr.dm [hunk 11]
```diff
@@ -2108,6 +2127,22 @@ var/global/list/belly_colorable_only_fullscreens = list("a_synth_flesh_mono",
 			unsaved_changes = TRUE
 			return TRUE
 		//CHOMPEdit end
+=======
+		if("set_vs_color")
+			if (istype(host, /mob/living/carbon/human))
+				var/mob/living/carbon/human/hhost = host
```
Comment (Verkister) [inline code/modules/vore/eating/vorepanel_vr.dm:2133]: why human exclusive?

File: code/modules/vore/eating/vorepanel_vr.dm (hunk 9 above)
Comment (Verkister) [reply code/modules/vore/eating/vorepanel_vr.dm:4132]: Also shouldn't be human exclusive.

File: code/modules/vore/eating/vorepanel_vr.dm [hunk 12]
```diff
@@ -292,20 +295,28 @@ var/global/list/belly_colorable_only_fullscreens = list("a_synth_flesh_mono",
 			"resist_animation" = selected.resist_triggers_animation,
 			"voresprite_size_factor" = selected.size_factor_for_sprite,
 			"belly_sprite_to_affect" = selected.belly_sprite_to_affect,
+<<<<<<< HEAD
 			"undergarment_chosen" = selected.undergarment_chosen,
 			"undergarment_if_none" = selected.undergarment_if_none || "None",
 			"undergarment_color" = selected.undergarment_color,
 			"belly_sprite_option_shown" = LAZYLEN(host.vore_icon_bellies) >= 1 ? TRUE : FALSE,
+=======
+			"belly_sprite_option_shown" = istype(host, /mob/living/carbon/human) ? (LAZYLEN(host:vore_icon_bellies) >= 1 ? TRUE : FALSE) : FALSE, // TODO: FIX THIS (It won't be fixed)
```
Comment (Verkister) [inline code/modules/vore/eating/vorepanel_vr.dm:304]: nuh uh not human exclusive.

File: code/modules/vore/eating/bellymodes_datum_vr.dm [hunk 13]
```diff
@@ -60,6 +61,19 @@ GLOBAL_LIST_INIT(digest_modes, list())
 			return
 
  		//CHOMPedit end
+=======
+				SEND_SOUND(L, sound(get_sfx("classic_death_sounds")))
+			else
+				SEND_SOUND(L, sound(get_sfx("fancy_death_prey")))
+		B.handle_digestion_death(L)
+		if(!L)
+			if (istype(B.owner, /mob/living/carbon/human))
```
Comment (Verkister) [inline code/modules/vore/eating/bellymodes_datum_vr.dm:70]: Not human exclusive. Same with line 106 below

File: code/modules/vore/eating/belly_obj_vr.dm [hunk 14]
```diff
@@ -676,6 +727,23 @@
...
+		L.clear_fullscreen("belly")
+		L.clear_fullscreen("belly2")
+		L.clear_fullscreen("belly3")
+		L.clear_fullscreen("belly4")
+		if(L.hud_used)
+			if(!L.hud_used.hud_shown)
+				L.toggle_hud_vis()
+		if((L.stat != DEAD) && L.ai_holder)
+			L.ai_holder.go_wake()
+	if (istype(owner, /mob/living/carbon/human))
```
Comment (Verkister) [inline code/modules/vore/eating/belly_obj_vr.dm:740]: This should not be in vore_fx

File: code/__defines/_planes+layers_vr.dm [hunk 15]
```diff
@@ -3,5 +3,6 @@
 #define PLANE_CH_STATUS_R			27 //Right-side status icon
 #define PLANE_CH_BACKUP				28 //Backup implant
 #define PLANE_CH_VANTAG				29 //Vore Antag hud
+#define PLANE_CH_STOMACH			30 //Stomach Plane
```
Comment (Verkister) [inline code/__defines/_planes+layers_vr.dm:6]: Ah. This explains the fucked up scaling mode thing. How many other layers will this cover? It's even above the hud icons. Are the bellysprites fullbright in the dark?

**Comment by SpadesNeil** [comment]
This PR broke Chomp bellies.
image
image

**Comment by Verkister** [comment]
Still need to fix the fuzzy/sharp scaling issue

File: code/modules/mob/living/carbon/human/update_icons.dm [hunk 16]
```diff
@@ -1447,6 +1447,74 @@ GLOBAL_LIST_EMPTY(damage_icon_parts) //see UpdateDamageIcon()
...
+		return
+
+	remove_layer(VORE_BELLY_LAYER)
+
+	var/image/vore_belly_image = get_vore_belly_image()
+	if(vore_belly_image)
+		vore_belly_image.layer = BODY_LAYER+VORE_BELLY_LAYER
+		overlays_standing[VORE_BELLY_LAYER] = vore_belly_image
+		vore_belly_image.plane = PLANE_CH_STOMACH //This one line of code. This ONE LINE OF CODE TOOK 6 HOURS TO FIGURE OUT. THANK YOU REDCAT.
+
```
Comment (Verkister) [inline code/modules/mob/living/carbon/human/update_icons.dm:1461]: vore_belly_image.appearance_flags = appearance_flags

**Review by Kashargul — APPROVED** [review]

**Comment by SpadesNeil** [comment]
Belly rub target is ignored when another player does it.

**Comment by Kashargul** [comment]
> Belly rub target is ignored when another player does it.

offtopic

**Pull request merged**